ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py

```diff
@@ -0,0 +1,154 @@
+
+    Args:
+        port: Port number to listen.
+        address: Endpoint address (default is localhost).
+        prefix: single-word application prefix relevant to the domain the metric belongs to.
+    """
+
+    def __init__(self, port: int = 8000, address: str = "", prefix: str = ""):
+        self._port = port
+        self._address = address
+        self._collector = CustomCollector(prefix)
+
+        start_http_server(port=self._port, addr=self._address)
+        REGISTRY.register(self._collector)
+
+    def export(
+        self, metric_records: Sequence[MetricRecord]
+    ) -> MetricsExportResult:
+        self._collector.add_metrics_data(metric_records)
+        return MetricsExportResult.SUCCESS
+
+    def shutdown(self) -> None:
+        REGISTRY.unregister(self._collector)
+
+
+class CustomCollector:
+    """ CustomCollector represents the Prometheus Collector object
+        https://github.com/prometheus/client_python#custom-collectors
+    """
+
+    def __init__(self, prefix: str = ""):
+        self._prefix = prefix
+        self._metrics_to_export = collections.deque()
+        self._non_letters_nor_digits_re = re.compile(
+            r"[^\w]", re.UNICODE | re.IGNORECASE
+        )
+
+    def add_metrics_data(self, metric_records: Sequence[MetricRecord]):
+        self._metrics_to_export.append(metric_records)
+
+    def collect(self):
+        """Collect fetches the metrics from OpenTelemetry
+        and delivers them as Prometheus Metrics.
+        Collect is invoked every time a prometheus.Gatherer is run
+        for example when the HTTP endpoint is invoked by Prometheus.
+        """
+
+        while self._metrics_to_export:
+            for metric_record in self._metrics_to_export.popleft():
+                prometheus_metric = self._translate_to_prometheus(
+                    metric_record
+                )
+                if prometheus_metric is not None:
+                    yield prometheus_metric
+
+    def _translate_to_prometheus(self, metric_record: MetricRecord):
+        prometheus_metric = None
+        label_values = []
+        label_keys = []
+        for label in metric_record.label_set.labels:
+            for label_tuple_value in label[1::2]:
+                label_values.append(label_tuple_value)
+
+        for label_key in metric_record.metric.label_keys:
+            label_keys.append(self._sanitize(label_key))
+        metric_name = ""
+        if self._prefix != "":
+            metric_name = self._prefix + "_"
+        metric_name += self._sanitize(metric_record.metric.name)
+
+        if isinstance(metric_record.metric, Counter):
+            prometheus_metric = CounterMetricFamily(
+                name=metric_name,
+                documentation=metric_record.metric.description,
+                labels=label_keys,
+            )
+            prometheus_metric.add_metric(
+                labels=label_values, value=metric_record.aggregator.checkpoint
```

Comment:
How does Prometheus handle discrepancies with missing label_keys/label values? Do they expect the same amount of keys and values passed in, and if so, we probably have to do some preprocessing here.

e.g1. label_keys: ["key1", "key2"], label_values: ["value1"]
eg2. label_keys: ["key1"], label_values: ["value1", "value2"]

Reply:
Prometheus will handle this correctly, if two labels are provided and only one value it will only add the populated one, same thing the other way around, I will update this logic to use MetricRecord.label_set.labels to be sure every single label key have a value

Reply:
Nice! Also seems like we might remove the recommended [keys](https://github.com/open-telemetry/opentelemetry-specification/issues/463) since Prometheus does not use this in their protocol.